[EuiMarkdownEditor] Added popover to display help syntax text when no custom plugins are available

miukimiu opened the pull request:
### Summary

Closes #5127.

- This PR adds a popover to display help syntax text when no custom plugins are available. 
- Adds an easier way to unregister plugins (thanks to @chandlerprall).
- Adds documentation to show how we can unregister plugins.

###  Design

If we're not showing syntax for plugins, we display the help syntax in a popover. No need to have a modal just with a link to the GitHub Flavored Markdown page.

<img width="1864" alt="md-no-plugin@2x" src="https://user-images.githubusercontent.com/2750668/134343352-f4bdd7ec-d4ea-4d47-85dc-5ad1402bc2a0.png">


The Markdown button was not displaying correctly on hover. 

<img width="313" alt="helper-btn@2x" src="https://user-images.githubusercontent.com/2750668/133259519-c99f5470-d720-4183-9d05-847c8676df15.png">


###  New docs section

As per my comment on https://github.com/elastic/eui/issues/5127#issuecomment-913701776. I think it isn't easy to know how to unregister plugins. For me, it makes sense to have a section in our docs to show how to do it. So I created the following section on `#/editors-syntax/markdown-editor#unregistering-plugins`. 

<img width="1262" alt="Screenshot 2021-09-14 at 13 38 27" src="https://user-images.githubusercontent.com/2750668/133259566-b5709651-11e7-4725-ba5d-dbdf1f9e0de6.png">

### Checklist

- [x] Check against **all themes** for compatibility in both light and dark modes
- ~[ ] Checked in **mobile**~
- [x] Checked in **Chrome**, **Safari**, **Edge**, and **Firefox**
- ~[ ] Props have proper **autodocs** and **[playground toggles](https://github.com/elastic/eui/blob/master/wiki/documentation-guidelines.md#adding-playground-toggles)**~
- [x] Added **[documentation](https://github.com/elastic/eui/blob/master/wiki/documentation-guidelines.md)**
- ~[ ] Checked **[Code Sandbox](https://codesandbox.io/)** works for the any docs examples~
- [x] Added or updated **[jest tests](https://github.com/elastic/eui/blob/master/wiki/testing.md)**
- ~[ ] Checked for **breaking changes** and labeled appropriately~
- ~[ ] Checked for **accessibility** including keyboard-only and screenreader modes~
- [x] A **[changelog](https://github.com/elastic/eui/blob/master/wiki/documentation-guidelines.md#changelog)** entry exists and is marked appropriately

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5147/

chandlerprall commented:
> I think it isn't easy to know how to unregister plugins. For me, it makes sense to have a section in our docs to show how to do it

🚀 great example idea; I've created a PR against your branch with an easier way to disable the tooltip plugin (and we can extend it with any other complete plugins in the future)

miukimiu commented:
@chandlerprall the CI is always failling 🤷🏽 with:

```js
15:25:16 Generating typescript definitions file
15:25:16 eui.d.ts(11728,29): error TS2307: Cannot find module 'unified/types/ts3.4/' or its corresponding type declarations.
15:26:16 eui.d.ts(11729,34): error TS2307: Cannot find module 'unified/types/ts3.4/' or its corresponding type declarations.
15:26:16 eui.d.ts(11729,377): error TS2307: Cannot find module 'unified/types/ts3.4/' or its corresponding type declarations.
15:26:16 child_process.js:669
15:26:16     throw err;
```

thompsongl commented:
👉 https://github.com/miukimiu/eui/pull/24 will fix the CI failure

miukimiu commented:
Thanks @thompsongl! 😄

miukimiu commented:
jenkins test this

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5147/

cchaos commented on CHANGELOG.md:
```diff
@@ -2,6 +2,11 @@
 
 - Added `placeholder` prop to `EuiMarkdownEditor` ([#5151](https://github.com/elastic/eui/pull/5151))
 - Added `.eui-textNumber` utility class to apply `tnum` font-feature setting ([#5078](https://github.com/elastic/eui/pull/5078))
+- Added a configuration parameter to the `EuiMarkdownEditor` plugin functions to exclude custom plugins ([#5147](https://github.com/elastic/eui/pull/5147))
```
Looks like you'll need to update the CL from master.

cchaos commented on src-docs/src/views/markdown_editor/mardown_editor_example.js:
```diff
@@ -62,6 +63,25 @@ const markdownEditorHeightSnippet = [
+import MarkdownEditorNoPlugins from './markdown_editor_no_plugins';
+const markdownEditorNoPluginsSource = require('!!raw-loader!./markdown_editor_no_plugins');
+const markdownEditorNoPluginsHtml = renderToHtml(MarkdownEditor);
+const markdownEditorNoPluginsSnippet = `
+const {
+  parsingPlugins,
+  processingPlugins,
+  uiPlugins,
+} = getDefaultEuiMarkdownPlugins({ exclude: ['tooltip'] });
+
+  <EuiMarkdownEditor
+    value={value}
+    onChange={setValue}
+    parsingPluginList={parsingPlugins}
+    processingPluginList={processingPlugins}
+    uiPlugins={uiPlugins}
+  />
+`;
```
```suggestion
const markdownEditorNoPluginsSnippet = `const {
  parsingPlugins,
  processingPlugins,
  uiPlugins,
} = getDefaultEuiMarkdownPlugins({ exclude: ['tooltip'] });

return (
  <EuiMarkdownEditor
    value={value}
    onChange={setValue}
    parsingPluginList={parsingPlugins}
    processingPluginList={processingPlugins}
    uiPlugins={uiPlugins}
  />
);`;
```

cchaos commented on src-docs/src/views/markdown_editor/mardown_editor_example.js:
```diff
@@ -108,6 +128,41 @@ export const MarkdownEditorExample = {
+          type: GuideSectionTypes.JS,
+          code: markdownEditorNoPluginsSource,
+        },
+        {
+          type: GuideSectionTypes.HTML,
+          code: markdownEditorNoPluginsHtml,
+        },
```
Reminder that we don't need to continue adding the `HTML` code config anymore.
```suggestion
```

cchaos commented on src-docs/src/views/markdown_editor/mardown_editor_example.js:
```diff
@@ -108,6 +128,41 @@ export const MarkdownEditorExample = {
+      title: 'Unregistering plugins',
+      text: (
+        <p>
+          Most of the built in plugins use the{' '}
+          <EuiLink href="https://github.github.com/gfm/" target="_blank">
+            GitHub flavored markdown syntax
+          </EuiLink>
+          . Other plugins like the tooltip don&apos;t follow this syntax, and
+          they might be unfamiliar or unnecessary in some contexts. For this
```
```suggestion
          The <strong>EuiMarkdownEditor</strong> comes with a default plugin
          for <EuiCode>tooltip</EuiCode> support. However, this may be unfamiliar
          or unnecessary in some contexts,
```

cchaos commented on src-docs/src/views/markdown_editor/mardown_editor_example.js:
```diff
@@ -108,6 +128,41 @@ export const MarkdownEditorExample = {
+          </EuiLink>
+          . Other plugins like the tooltip don&apos;t follow this syntax, and
+          they might be unfamiliar or unnecessary in some contexts. For this
+          reason, you can unregister the plugin by excluding it from the
+          <EuiCode>parsingPlugins</EuiCode>,{' '}
+          <EuiCode>processingPlugins</EuiCode> and <EuiCode>uiPlugins</EuiCode>{' '}
+          list. This way, you make sure the syntax won&apos;t be identified and
```
```suggestion
          and you can unregister this plugin by excluding it from the
          <EuiCode>parsingPlugins</EuiCode>,{' '}
          <EuiCode>processingPlugins</EuiCode> and <EuiCode>uiPlugins</EuiCode>{' '}
          options with a single <EuiCode>exclude</EuiCode> parameter passed to <EuiCode>getDefaultEuiMarkdownPlugins()<EuiCode>. This will ensure the syntax won&apos;t be identified or
```

cchaos commented on src-docs/src/views/markdown_editor/mardown_editor_example.js:
```diff
@@ -108,6 +128,41 @@ export const MarkdownEditorExample = {
+          <EuiCode>parsingPlugins</EuiCode>,{' '}
+          <EuiCode>processingPlugins</EuiCode> and <EuiCode>uiPlugins</EuiCode>{' '}
+          list. This way, you make sure the syntax won&apos;t be identified and
+          consequently rendered. No additional UI, like the button and help
+          syntax, will be displayed.
+        </p>
```
```suggestion
          rendered and no additional UI, like the button and help
          syntax, will be displayed.
        </p>
```

cchaos commented on src-docs/src/views/markdown_editor/markdown_editor_no_plugins.js:
```diff
@@ -0,0 +1,83 @@
+!{tooltip[anchor text](Tooltip content)}
+`;
+
+const dropHandlers = [
+  {
+    supportedFiles: ['.jpg', '.jpeg'],
+    accepts: (itemType) => itemType === 'image/jpeg',
```
I think it would be good to exclude this file handler as well as this whole section is mainly talking about a plain and simple editor.

cchaos commented on src-docs/src/views/markdown_editor/markdown_editor_no_plugins.js:
```diff
@@ -0,0 +1,83 @@
+        aria-label="EUI markdown editor with no default plugins demo"
+        value={value}
+        onChange={setValue}
+        height={400}
```
Again, just trying to reduce this example to simply what is needed.
```suggestion
```

cchaos commented on src-docs/src/views/markdown_editor/markdown_editor_no_plugins.js:
```diff
@@ -0,0 +1,83 @@
+          {isAstShowing ? 'Hide editor AST' : 'Show editor AST'}
+        </EuiButton>
+      </div>
+      {isAstShowing && <EuiCodeBlock language="json">{ast}</EuiCodeBlock>}
```
**Can be in a follow up PR:**

Because this continues to be added to every examples, and I can understand the nicety to it, I start worrying that it actually looks like it comes with the component. Can we rethink a way to render this. Like maybe look at how the Page examples are wrapped in a custom component but only renders what it needed for the consumer.

miukimiu replied:
I'll follow up on this. 👍🏽

cchaos commented on src/components/markdown_editor/_markdown_editor_footer.scss:
```diff
@@ -34,8 +34,9 @@
-.euiMarkdownEditorFooter__help {
+.euiMarkdownEditorFooter__helpButton {
   justify-self: flex-end;
+  min-width: $euiSizeXL;
```
My suggestion here would be to actually make the `EuiButtonIcon` `size="s"` instead of `xs`. This will give you the same size in width, a little taller, but that's ok given that it gives it more clickable area.

cchaos commented on src/components/markdown_editor/markdown_editor_footer.tsx:
```diff
@@ -173,82 +175,128 @@ export const EuiMarkdownEditorFooter = forwardRef<
+  const hasUiPlugins = uiPlugins.length > 0;
+  const githubSyntaxLink = (
+    <EuiLink href="https://github.github.com/gfm/" target="_blank">
+      GitHub flavored markdown
```
Needs i18n

cchaos commented on src/components/markdown_editor/markdown_editor_footer.tsx:
```diff
@@ -173,82 +175,128 @@ export const EuiMarkdownEditorFooter = forwardRef<
+          <EuiModal onClose={() => setIsShowingHelpModal(false)}>
+            <EuiModalHeader>
+              <EuiTitle>
+                <h3>
```
Is the `h3` correct here? My assumption with modals is that they're presented as their own doc structure so the titles must always be an `h1`.

cchaos commented on src/components/markdown_editor/markdown_editor_footer.tsx:
```diff
@@ -173,82 +175,128 @@ export const EuiMarkdownEditorFooter = forwardRef<
-          </EuiModalFooter>
-        </EuiModal>
-      )}
+      {helpSyntaxButton}
```
Thoughts on wrapping this in a tooltip with "Syntax help" (same as the popover title). There's currently no feedback to the user what this button will do on hover.

cchaos reviewed (changes requested):
I like the simplicity of just showing the popover when there's not much more context than just pointing to GH markdown docs. 💯 

The component fixes mostly LGTM (with a few fixes mentioned), but most of my comments revolve around the docs to try to make it clearer / get straight to the point.

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5147/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5147/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5147/

miukimiu commented:
@cchaos I finished the changes based on your review:

- No more HTML code.
- I wrapped the markdown button with a tooltip.
- Updated the markdown button to size "S" and updated the upload button to match the same size.
- Simplified the example and removed unnecessary code like upload button, errors, and show editor AST.

thompsongl reviewed (dismissed):
Thanks, @miukimiu!

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5147/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5147/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5147/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5147/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5147/

miukimiu commented:
@cchaos and @thompsongl these are the new changes:

- The popover is back when no plugins are available. For this scenario, I'm using the button title with "Syntax help" instead of a tooltip so it doesn't conflict with the open popover.
- Linking the "GitHub flavored markdown" to https://guides.github.com/features/mastering-markdown/. The other page we were linking before was too complex and we were not supporting a lot of things. @cchaos we talked about linking to https://www.markdownguide.org/ instead but I came to the conclusion that we're following more the  GitHub flavored markdown. We support emojis, task lists, and strikethrough.

cchaos commented on src/components/markdown_editor/markdown_editor_footer.tsx:
```diff
@@ -173,82 +183,134 @@ export const EuiMarkdownEditorFooter = forwardRef<
     );
   }
 
+  const hasUiPlugins = uiPlugins.length > 0;
```
I noticed that in the modal content the list of plugins only show if the plugin has `helpText` =(`uiPlugins.filter(({ helpText }) => !!helpText)`). But that filter isn't applied until **inside** the modal content. I'm thinking you might want to pull that up outside fo the if statement otherwise you could end up with an empty modal.

miukimiu replied:
Good catch! Done! 👍🏽

cchaos reviewed (commented):
👍 The UI looks great. But I did find a possible edge case where if there are plugins but none of the plugins actually have `helpText`.

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5147/

cchaos reviewed (approved):
👍 Double-checked the different examples with the new logic. LGTM!

Rebase and CL update needed.

miukimiu commented:
> Rebase and CL update needed.

Thanks, @cchaos! I confirmed and i's updated and rebased. 
@thompsongl I'm merging it because I reverted the code to what it was when you approved.
